cmd/zrepl/zrepl.c

```diff
@@ -833,6 +870,56 @@ uzfs_zvol_io_conn_acceptor(void)
 	zk_thread_exit();
 }
 
+/*
+ * This func takes care of sending potentially multiple read blocks each
+ * prefixed by metainfo.
+ */
+static int
+uzfs_send_reads(int fd, zvol_io_cmd_t *zio_cmd)
+{
+	zvol_io_hdr_t 	*hdr = &zio_cmd->hdr;
+	struct zvol_io_rw_hdr read_hdr;
+	metadata_desc_t	*md;
+	int	rc = 0;
+
+	/* special case for missing metadata */
+	if (zio_cmd->metadata_desc == NULL) {
+		read_hdr.io_num = 0;
+		read_hdr.len = hdr->len;
+		rc = uzfs_zvol_socket_write(fd, (char *)&read_hdr,
+		    sizeof (read_hdr));
+		if (rc != 0)
+			return (rc);
+		rc = uzfs_zvol_socket_write(fd, zio_cmd->buf, hdr->len);
+		return (rc);
+	}
+
+	for (md = zio_cmd->metadata_desc; md != NULL; md = md->next) {
+		read_hdr.io_num = md->metadata.io_num;
+		read_hdr.len = md->len;
+		rc = uzfs_zvol_socket_write(fd, (char *)&read_hdr,
+		    sizeof (read_hdr));
+		if (rc != 0)
+			goto end;
+
+		rc = uzfs_zvol_socket_write(fd,
+		    (char *)zio_cmd->buf + md->rel_offset, md->len);
```

Comment:
rel_offset may not be required. this can be calculated by adding (md->len) of the data wrote till now

Reply:
correct. I will remove it.